ocaml/xenopsd/c_stubs/xenctrlext_stubs.c

```diff
@@ -394,6 +409,127 @@ CAMLprim value stub_xenctrlext_cputopoinfo(value xch)
 	CAMLreturn(result);
 }
 
+CAMLprim value stub_xenforeignmemory_open(value logger, value open_flags)
+{
+        CAMLparam2(logger, open_flags);
+        struct xentoollog_logger *log_handle = NULL;
+        struct xenforeignmemory_handle *fmem;
+        CAMLlocal1(result);
+
+        if(Is_some(logger)) {
+                log_handle = Xtl_val(Some_val(logger));
+        }
+
+        fmem = xenforeignmemory_open(log_handle, Int_val(open_flags));
+
+        if(fmem == NULL) {
+                caml_failwith("Error when opening foreign memory handle");
+        }
+
+        result = caml_alloc(1, Abstract_tag);
+        Xfm_val(result) = fmem;
+
+        CAMLreturn(result);
+}
+
+CAMLprim value stub_xenforeignmemory_close(value fmem)
+{
+        CAMLparam1(fmem);
+        int retval;
+
+        retval = xenforeignmemory_close(Xfm_val(fmem));
+
+        if(retval < 0) {
+                caml_failwith("Error when closing foreign memory handle");
+        }
+
+        CAMLreturn(Val_unit);
+}
+
+CAMLprim value stub_xenforeignmemory_map(value fmem, value dom,
+        value prot_flags, value pages)
+{
+        CAMLparam4(fmem, dom, prot_flags, pages);
+        CAMLlocal2(cell, result);
+        size_t i, pages_length;
+        xen_pfn_t *arr;
+        int prot, the_errno;
+        void *retval;
+
+        if (Field(prot_flags, 0) == Val_false &&
+            Field(prot_flags, 1) == Val_false &&
+            Field(prot_flags, 2) == Val_false) {
+                prot = PROT_NONE;
+        } else {
+                prot = 0;
+                if(Field(prot_flags, 0) == Val_true) {
+                        prot |= PROT_READ;
+                }
+                if(Field(prot_flags, 1) == Val_true) {
+                        prot |= PROT_WRITE;
+                }
+                if(Field(prot_flags, 2) == Val_true) {
+                        prot |= PROT_EXEC;
+                }
+        }
+
+        // traverse list to know the length of the array
+        cell = pages;
+        for(pages_length = 0; cell != Val_emptylist; pages_length++) {
+                cell = Field(cell, 1);
+        }
+
+        // allocate and populate the array
+        arr = malloc(sizeof(xen_pfn_t) * pages_length);
+        if(arr == NULL) {
+                caml_failwith("Error: could not allocate page array before mapping memory");
+        }
+
+        cell = pages;
+        for(i = 0; i < pages_length; i++) {
+                arr[i] = Int64_val(Field(cell, 0));
+                cell = Field(cell, 1);
+        }
+
+        retval = xenforeignmemory_map
+                (Xfm_val(fmem), _D(dom), prot, pages_length, arr, NULL);
+        the_errno = errno;
+
+        free(arr);
+
+        if(retval == NULL) {
+                raise_unix_errno_msg(the_errno,
+                                "Error when trying to map foreign memory");
+        }
+
+        result = caml_ba_alloc_dims(
```

Comment:
I don't know how this works in detail - but this call must no longer depend on `arr` because it was freed.

Reply:
Indeed, `arr` is only used to feed the pages to Xen's function `xenforeignmemory_map`. It's not used here to allocate the bigarray that is being returned